File: Java/src/_1600_Throne_Inheritance/Solution.java
package _1600_Throne_Inheritance;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

/**
 * Date: 2021/6/20 10:32
 */
public class Solution {

    static class ThroneInheritance {

        HashMap<String, List<String>> edges = new HashMap<>();
        HashSet<String> dead = new HashSet<>();
        String king;

        public ThroneInheritance(String kingName) {
            king = kingName;
        }

        public void birth(String p, String c) {
            List<String> children = edges.getOrDefault(p, new ArrayList<>());
            children.add(c);
            edges.put(p, children);
        }

        public void death(String name) {
            dead.add(name);
        }

        public List<String> getInheritanceOrder() {
            List<String> ans = new ArrayList<>();
            preorder(ans, king);
            return ans;
        }

        void preorder(List<String> ans, String name) {
            if (!dead.contains(name)) ans.add(name);
            for (String c : edges.getOrDefault(name, new ArrayList<>())) preorder(ans, c);
        }
    }

    public static void main(String[] args) {

        ThroneInheritance t= new ThroneInheritance("king");  // 继承顺序：king
        t.birth("king", "andy");  // 继承顺序：king > andy
        t.birth("king", "bob");  // 继承顺序：king > andy > bob
        t.birth("king", "catherine");  // 继承顺序：king > andy > bob > catherine
        t.birth("andy", "matthew");  // 继承顺序：king > andy > matthew > bob > catherine
        t.birth("bob", "alex");  // 继承顺序：king > andy > matthew > bob > alex > catherine
        t.birth("bob", "asha");  // 继承顺序：king > andy > matthew > bob > alex > asha > catherine
        System.out.println(t.getInheritanceOrder());  // [king, andy, matthew, bob, alex, asha, catherine]
        t.death("bob");  // 继承顺序：king > andy > matthew > bob（已经去世）> alex > asha > catherine
        System.out.println(t.getInheritanceOrder());  // [king, andy, matthew, alex, asha, catherine]
    }
}

File: Python/_1600_Throne_Inheritance/Solution.py
# coding=utf-8
# Date: 2021/6/20 10:52
from typing import List


# 执行用时：504 ms, 在所有 Python3 提交中击败了87.50%的用户
# 内存消耗：64.1 MB, 在所有 Python3 提交中击败了67.86%的用户
class ThroneInheritance:

    def __init__(self, kingName: str):
        self.edges = {}
        self.dead = set()
        self.king = kingName

    def birth(self, p: str, c: str) -> None:
        if p in self.edges:
            self.edges[p].append(c)
        else:
            self.edges[p] = [c]

    def death(self, name: str) -> None:
        self.dead.add(name)

    def getInheritanceOrder(self) -> List[str]:
        ans = []
        self.preorder(ans, self.king)
        return ans

    def preorder(self, ans, name):
        if name not in self.dead:
            ans.append(name)
        if name in self.edges:
            for c in self.edges[name]:
                self.preorder(ans, c)


if __name__ == "__main__":
    t = ThroneInheritance("king")
    t.birth("king", "andy")
    t.birth("king", "bob")
    t.birth("king", "catherine")
    t.birth("andy", "matthew")
    t.birth("bob", "alex")
    t.birth("bob", "asha")
    print(t.getInheritanceOrder())  # ['king', 'andy', 'matthew', 'bob', 'alex', 'asha', 'catherine']
    t.death("bob")
    print(t.getInheritanceOrder())  # ['king', 'andy', 'matthew', 'alex', 'asha', 'catherine']
